Sage/carbon
# fix(dialog, sidebar): use aria-hidden and inert to hide non-modal elements

Although we use the aria-modal attribute on the dialog content, this does not fully hide the background content from all assistive technologies - notably speech recognition tools. This PR introduces a new custom hook that adds the `aria-hidden="true"` and `inert` attributes to all containers of non-modal content, to further mark them as non-active for assistive technologies. The attributes are removed - or returned to any previous values they had - when the modal is closed. The new hook also determines whether a modal is the top modal or not, which was previously done inside each individual component, in order to centralise and simplify this logic.

fix #5920

### Proposed behaviour



When using speech recognition technology, when a modal component (dialog or sidebar) is open, saying "show numbers" should not cause numbers to be assigned to any elements on the page that are not inside the modal, as they are not able to be interacted with at that time. 

### Current behaviour



With Windows speech recognition, saying "show numbers" causes elements that are outisde the modal to be assigned numbers (eg the button which opened the dialog which is currently open).

### Checklist




#### QA


### Additional context



### Testing instructions



The PR can be tested with the sandbox linked in the Github issue, or on Storybook with any of our example stories for Dialog, DialogFullScreen, and Sidebar.

Test with any available speech recognition technologies, particularly:
- Voice Control on Mac (ideally with Safari)
- Windows Speech Recognition on Windows (in all supported browsers - particularly Chrome, Edge and Firefox)
- any other speech recognition tools that are available to test with.
In each, test that asking to "show numbers" with the dialog/sidebar open does not assign numbers to any otherwise-interactive elements outside the modal component (particularly any button that opens the modal). Check that numbers are assigned to these again after closing the dialog.
Ensure there are no regressions with such speech recognition tools - both with "show numbers" and more generally around modals, anything that works OK on master should still work OK on the PR!
Please inform me of any browser/OS/assistive-tech combinations where the fix does not work - however provided that it works on some and does not make the situation worse on any then this should probably not block the PR!

Also test with all supported screenreader/browser combinations to ensure there are no regressions with assistive technologies that are not connected to speech recognition.
